### PR TITLE
feat: support local volume (data path)

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/3957

Signed-off-by: Derek Su [derek.su@suse.com](mailto:derek.su@suse.com)

### CODE REVIEW

File: app/cmd/replica.go [hunk 1]
```diff
@@ -99,7 +109,12 @@ func startReplica(c *cli.Context) error {
 		}
 	}
 
-	controlAddress, dataAddress, syncAddress, syncPort, err := util.ParseAddresses(address)
+	volumeName := c.String("volume-name")
+
```
Comment (innobead) [inline app/cmd/replica.go:113]: No need needless empty line.

**Review by innobead — DISMISSED** [review]
LGTM

File: pkg/replica/rpc/dataserver.go [hunk 2]
```diff
@@ -47,3 +61,28 @@ func (s *DataServer) ListenAndServe() error {
...
+
+	l, err := net.ListenUnix("unix", unixAddr)
+	if err != nil {
+		return err
+	}
+
+	for {
+		conn, err := l.AcceptUnix()
+		if err != nil {
+			logrus.Errorf("Failed to accept unix-domain-socket connection since %v", err)
```
Comment (c3y1huang) [inline pkg/replica/rpc/dataserver.go:79]:
```suggestion
			logrus.WithError(err).Error("failed to accept unix-domain-socket connection")
```

File: pkg/replica/rpc/dataserver.go [hunk 3]
```diff
@@ -10,18 +11,31 @@ import (
 )
 
 type DataServer struct {
-	address string
-	s       *replica.Server
+	protocol string
```
Comment (c3y1huang) [inline pkg/replica/rpc/dataserver.go:14]: Why not use `types.DataServerProtocol`?

File: app/cmd/replica.go (hunk 1 above)
Comment (derekbit) [reply app/cmd/replica.go]: Removed. Thank you.

File: pkg/replica/rpc/dataserver.go [hunk 4]
```diff
@@ -35,7 +49,7 @@ func (s *DataServer) ListenAndServe() error {
 	for {
 		conn, err := l.AcceptTCP()
 		if err != nil {
-			logrus.Errorf("failed to accept connection %v", err)
+			logrus.Errorf("Failed to accept tcp connection since %v", err)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/replica/rpc/dataserver.go]: Updated, thank you.

File: pkg/replica/rpc/dataserver.go (hunk 3 above)
Comment (derekbit) [reply pkg/replica/rpc/dataserver.go]: Updated, thank you.

**Review by innobead — APPROVED** [review]

File: pkg/replica/rpc/dataserver.go [hunk 5]
```diff
@@ -1,27 +1,42 @@
...
-		s:       s,
+		protocol: protocol,
+		address:  address,
+		s:        s,
 	}
 }
 
 func (s *DataServer) ListenAndServe() error {
+	switch s.protocol {
+	case "tcp":
```
Comment (c3y1huang) [inline pkg/replica/rpc/dataserver.go:30]:
```suggestion
	case types.DataServerProtocolTCP:
```

File: pkg/replica/rpc/dataserver.go [hunk 6]
```diff
@@ -1,27 +1,42 @@
...
+		return s.listenAndServeTCP()
+	case "unix":
```
Comment (c3y1huang) [inline pkg/replica/rpc/dataserver.go:32]:
```suggestion
	case types.DataServerProtocolUNIX:
```

File: pkg/util/util.go [hunk 7]
```diff
@@ -331,3 +334,16 @@ func GetFunctionName(i interface{}) string {
 func RandomID(randomIDLenth int) string {
 	return UUID()[:randomIDLenth]
 }
+
+func GetAddresses(volumeName, address string, dataServerProtocol types.DataServerProtocol) (string, string, string, int, error) {
+	switch dataServerProtocol {
+	case "tcp":
```
Comment (c3y1huang) [inline pkg/util/util.go:340]:
```suggestion
	case types.DataServerProtocolTCP:
```

File: pkg/util/util.go [hunk 8]
```diff
@@ -331,3 +334,16 @@ func GetFunctionName(i interface{}) string {
...
+		return ParseAddresses(address)
+	case "unix":
```
Comment (c3y1huang) [inline pkg/util/util.go:342]:
```suggestion
	case types.DataServerProtocolUNIX:
```

File: pkg/replica/rpc/dataserver.go (hunk 6 above)
Comment (derekbit) [reply pkg/replica/rpc/dataserver.go]: Updated. Thank you.

File: pkg/util/util.go (hunk 7 above)
Comment (derekbit) [reply pkg/util/util.go]: Updated. Thank you.

File: pkg/util/util.go (hunk 8 above)
Comment (derekbit) [reply pkg/util/util.go]: Updated. Thank you.

File: pkg/replica/rpc/dataserver.go (hunk 5 above)
Comment (derekbit) [reply pkg/replica/rpc/dataserver.go]: Updated. Thank you.

**Review by c3y1huang — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Pull request merged**